pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,474 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_primitives::{SECONDS_PER_MONTH, SECONDS_PER_YEAR};
+use cfg_traits::{interest::InterestRate, Seconds};
+use chrono::{DateTime, Datelike, Days, Months, NaiveDate, TimeDelta};
+use chronoutil::DateRule;
+use frame_support::pallet_prelude::RuntimeDebug;
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{
+		EnsureAddAssign, EnsureDiv, EnsureFixedPointNumber, EnsureInto, EnsureSub, EnsureSubAssign,
+	},
+	ArithmeticError, DispatchError, FixedPointNumber, FixedPointOperand,
+};
+
+/// Specify the expected repayments date
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum Maturity {
+	/// Fixed point in time, in secs
+	Fixed {
+		/// Secs when maturity ends
+		date: Seconds,
+		/// Extension in secs, without special permissions
+		extension: Seconds,
+	},
+}
+
+impl Maturity {
+	pub fn fixed(date: Seconds) -> Self {
+		Self::Fixed { date, extension: 0 }
+	}
+
+	pub fn date(&self) -> Seconds {
+		match self {
+			Maturity::Fixed { date, .. } => *date,
+		}
+	}
+
+	pub fn is_valid(&self, now: Seconds) -> bool {
+		match self {
+			Maturity::Fixed { date, .. } => *date > now,
+		}
+	}
+
+	pub fn extends(&mut self, value: Seconds) -> Result<(), ArithmeticError> {
+		match self {
+			Maturity::Fixed { date, extension } => {
+				date.ensure_add_assign(value)?;
+				extension.ensure_sub_assign(value)
+			}
+		}
+	}
+}
+
+fn seconds_to_days(seconds: Seconds) -> Result<u32, DispatchError> {
+	Ok(TimeDelta::try_seconds(seconds.ensure_into()?)
+		.ok_or(DispatchError::Other("Precission error with seconds"))?
+		.num_days()
+		.ensure_into()?)
+}
+
+fn seconds_to_date(date_in_seconds: Seconds) -> Result<NaiveDate, DispatchError> {
+	Ok(DateTime::from_timestamp(date_in_seconds.ensure_into()?, 0)
+		.ok_or(DispatchError::Other("Invalid date in seconds"))?
+		.date_naive())
+}
+
+fn date_to_seconds(date: NaiveDate) -> Result<Seconds, DispatchError> {
+	Ok(date
+		.and_hms_opt(0, 0, 0)
+		.ok_or(DispatchError::Other("Invalid h/m/s"))?
+		.and_utc()
+		.timestamp()
+		.ensure_into()?)
+}
+
+fn seconds_in_month(date: NaiveDate) -> Result<Seconds, DispatchError> {
+	todo!()
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum ReferenceDate {
+	/// Payments are expected every period relative to a specific date.
+	/// E.g. if the period is monthly and the specific date is Mar 3, the
+	/// first interest payment is expected on Apr 3.
+	Date(Seconds),
+
+	/// At the end of the period, e.g. the last day of the month for a monthly
+	/// period
+	End,
+}
+
+impl ReferenceDate {
+	fn monthly_cashflow_dates(
+		&self,
+		start: Seconds,
+		end: Seconds,
+	) -> Result<Vec<(Seconds, Seconds)>, DispatchError> {
```

Comment:
What is the `(Seconds, Seconds)` representing?

Reply:
Still WIP, the first is the date in seconds when you need to make the payment, and the second is the interval for that payment. Normally, the interval will be the whole month, except for the first and last months that it can be fewer days

Reply:
And we need the interval for calculation purposes?

Reply:
Yes, we would need it later to apply a proportional interest for each month

Reply:
Suppose you pay at the end of the month, but you've borrowed on day 15. The first month you will pay only half of the month. The same will happen with the last month.